signer/threshold_validator.go

```diff
@@ -103,12 +103,90 @@ func (pv *ThresholdValidator) mustLoadChainState(chainID string) ChainSignState
 // sign process if it is greater than the current high watermark. A mutex is used to avoid concurrent
 // state updates. The disk write is scheduled in a separate goroutine which will perform an atomic write.
 // pendingDiskWG is used upon termination in pendingDiskWG to ensure all writes have completed.
-func (pv *ThresholdValidator) SaveLastSignedStateInitiated(chainID string, signState SignStateConsensus) error {
+func (pv *ThresholdValidator) SaveLastSignedStateInitiated(chainID string, block *Block) ([]byte, time.Time, error) {
 	css := pv.mustLoadChainState(chainID)
 
-	css.lastSignStateInitiatedMutex.Lock()
-	defer css.lastSignStateInitiatedMutex.Unlock()
-	return css.lastSignStateInitiated.Save(signState, &pv.pendingDiskWG)
+	height, round, step := block.Height, block.Round, block.Step
+
+	err := css.lastSignStateInitiated.Save(NewSignStateConsensus(height, round, step), &pv.pendingDiskWG)
+	if err == nil {
+		// good to sign
+		return nil, time.Time{}, nil
+	}
+
+	existingSignature, existingTimestamp, sameBlockErr := pv.getExistingBlockSignature(chainID, block)
+
+	if _, ok := err.(*SameHRSError); ok {
+		// Wait for last sign state signature to be the same block
+		if sameBlockErr == nil {
+			if existingSignature != nil {
+				return existingSignature, existingTimestamp, nil
+			}
+			// good to sign again
+			return nil, time.Time{}, nil
+		}
+
+		if _, ok := sameBlockErr.(*StillWaitingForBlockError); !ok {
+			return nil, existingTimestamp, sameBlockErr
+		}
+
+		// the cluster is already in progress signing this block
+		// wait for cluster to finish before proceeding
+
+		css.lastSignState.cond.L.Lock()
+		defer css.lastSignState.cond.L.Unlock()
+		for {
+			css.lastSignState.cond.Wait()
+
+			ssc, ok := css.lastSignState.cache[block.HRSKey()]
+			if !ok {
+				pv.logger.Debug(
+					"Cond does not exist in cache",
```

Comment:
Is it actually `Cond` that doesn't exist in the cache here?

Maybe the log message should note that it is waiting for the next sign state signal, for a bit more detail for the human log reader.